### PR TITLE
[OError] getFullInfo returns the error context of nested causes

### PR DESCRIPTION
**Pull request by das7pad**
For https://github.com/overleaf/issues/issues/3651

This PR is extending `OError.getFullInfo` in merging the error context of nested causes.

As per https://github.com/overleaf/issues/issues/3792 no internal context is leaked to the user.

---

Deployment: squash merge.

### CODE REVIEW

**Review by jdleesmiller — COMMENTED** [review]
The implementation all looks good 👍 

Could you elaborate on why it's needed for the referenced issue? I'm not sure I'm seeing the connection.

I think it would also help to have some clearer guidance on which method to use. Are we moving toward changing `getFullInfo` in some future version to use the new behaviour (include causes)? If so do we want to deprecate the old method (temporarily?)?

I am leaning toward making `getFullInfo` include causes, with the rational that `info` should be for logging. A custom error class would be more appropriate if there are properties to be made public / reasoned about.

**Comment by das7pad** [comment]
> The implementation all looks good 
> 
> Could you elaborate on why it's needed for the referenced issue? I'm not sure I'm seeing the connection.

Thanks, good questions!

The SmokeTests shall use chained errors. Lower errors may attach additional context in the `.info` field.
For example a `/project/someId` request may return a 404 response code, but an assertion expects a 200. In the context of the assertion error I want to put the expected and actual code. Outside of the test case (I call them `step` in the rewrite), the assertion error is catched and functions as cause of a generic message like `loadEditor failed` (which in turn is _public_). For the logging statement I would like to include the context in a serialized format.

At the tip of https://github.com/overleaf/web-internal/commit/634a52ebdbdabdd3888b0de963778753bb24e4a4 log messages for failed smoke tests look like this now:

<details>

(This is with a [_patched_ error serializer for the logger](https://github.com/overleaf/logger-module/blob/4355a257b0d65180d0e9e647430d15d0627ea9d8/logging-manager.js#L13-L20). And we may want to move the `cause` into a separate field too.) 
``` 
web$ MOCHA_GREP=SmokeTests make test_acceptance_app
...
{
  "name": "web",
  "hostname": "b9c8b6c880ab",
  "pid": 30,
  "level": 50,
  "err": {
    "message": "run.101_loadEditor failed",
    "name": "SmokeTestFailure",
    "stack": "SmokeTestFailure: run.101_loadEditor failed\n    at runAndTrack (/app/test/smoke/src/SmokeTests.js:50:13)\n    at process._tickCallback (internal/process/next_tick.js:68:7)\ncaused by:\n    OError: unexpected response code\n        at assertHasStatusCode (/app/test/smoke/src/support/requestHelper.js:49:11)\n        at run (/app/test/smoke/src/steps/101_loadEditor.js:6:3)\n        at process._tickCallback (internal/process/next_tick.js:68:7)",
    "info": {
      "stats": {
        "start": "2020-11-27T17:48:28.228Z",
        "steps": [
          {
            "init": 0
          },
          {
            "run.000_getLoginCsrf": 6
          },
          {
            "run.001_clearRateLimits": 0
          },
          {
            "run.002_login": 9
          },
          {
            "run.100_loadProjectDashboard": 26
          },
          {
            "run.101_loadEditor": 108
          },
          {
            "cleanup.002_login": 10
          }
        ],
        "end": "2020-11-27T17:48:28.387Z",
        "duration": 159
      },
      "actual": 404,
      "expected": 200
    }
  },
  "msg": "health check failed",
  "time": "2020-11-27T17:48:28.388Z",
  "v": 0
}
```
</details>

> I think it would also help to have some clearer guidance on which method to use. Are we moving toward changing `getFullInfo` in some future version to use the new behaviour (include causes)? If so do we want to deprecate the old method (temporarily?)?

For the moment I would like to them both available. A followup ticket could look into the usage of error causes and the use of potentially sensible details that should not end up in logs. When these are cleared we can merge `getFullInfo` and `getFullInfoIncludeCause` (or deprecate `getFullInfo`).

> I am leaning toward making `getFullInfo` include causes, with the rational that `info` should be for logging. A custom error class would be more appropriate if there are properties to be made public / reasoned about. [And this could be a good step in that direction.]

For the record: I do not want to expose error context of the tests to the public, it's only used for logging.

`dev-env $ grep -r . -e getFullInfo` shows [one usage](https://github.com/overleaf/web-internal/blob/62dbd41da9f5dbb11ddd26bc19775b4e9a6432ca/app/src/Features/Subscription/SubscriptionController.js#L233) in production code [(and logging)](https://github.com/overleaf/logger-module/blob/4355a257b0d65180d0e9e647430d15d0627ea9d8/logging-manager.js#L13-L20). There are a few other hits in web tests.
The first hit needs looking into in case we merge the two methods.

**Review by emcsween — COMMENTED** [review]
I'm also worried about having two slightly different ways to do the same thing. As a user of the library, I don't think I'd have a good way to choose which method to use, so we can just expect random use of one or the other, and when wrapping errors, I wouldn't know whether to expect that the info from the underlying error would get passed along or not.

Looking at your smoke tests branch, the only reason to use causes rather than tags is to [test whether an error is a smoke test failure or another type of error](https://github.com/overleaf/web-internal/blob/87ff192e6f962a2bda19892530f6dc6bcf051ecf/app/src/Features/HealthCheck/HealthCheckController.js#L106-L108), and replace the log message by "low level error" when the error is not an expected smoke test failure. I would argue that the original error message actually carries more information and it would be fine to use `OError.tag()` everywhere, which would achieve the same effect of making the info from the original error visible in logs.

If we nevertheless want to wrap errors, it's possible to expose the info from the cause one level deep in a simple way without adding another method to OError. Instead of `OError.getFullInfoIncludeCause(err)`, you could write `{ OError.getFullInfo(err.cause), ...info }` or even: `{ ...info, cause: OError.getFullInfo(err.cause) }`. Or, you could do something similar in the `SmokeTestFailure` constructor and manually initialize the `info` object with the info from the underlying error. I think it would make it clearer that we're doing something different from the usual behaviour.

Finally, if we think it's actually better if info from causes is passed along to the wrapping error, then I'd much prefer it if `getFullInfo()` had its behaviour changed.

**Comment by das7pad** [comment]
Thanks for the feedback, Eric.

> I'm also worried about having two slightly different ways to do the same thing. As a user of the library, I don't think I'd have a good way to choose which method to use, so we can just expect random use of one or the other, and when wrapping errors, I wouldn't know whether to expect that the info from the underlying error would get passed along or not.

Indeed, this is a tricky problem. There are cases to be made about not exposing internals of causes, and also about returning all the details in a chain of errors.

WDYT about Johns proposal of a custom OError class that allows nested context to get returned? (from `OError.getFullInfo` I assume)

> Looking at your smoke tests branch, the only reason to use causes rather than tags is to [test whether an error is a smoke test failure or another type of error](https://github.com/overleaf/web-internal/blob/87ff192e6f962a2bda19892530f6dc6bcf051ecf/app/src/Features/HealthCheck/HealthCheckController.js#L106-L108), and replace the log message by "low level error" when the error is not an expected smoke test failure. 

Using OError tags does not work for ioredis related error as it's reusing error instances. Hence I do not think it's a good idea to tag errors on the framework level which may clobber context in the app.
The idea of the `SmokeTestFailure` error class is having a very limited amount of info in it's message, that in turn is OK to expose in the response payload.

> I would argue that the original error message actually carries more information and it would be fine to use `OError.tag()` everywhere, which would achieve the same effect of making the info from the original error visible in logs.

`ioredis` and `request` errors can be very cryptic and they can also include internals that should not get exposed (e.g. internal ip addresses). I think a message describing the error and linking to the actual error via .cause is a better approach in this case.
E.g. for the cleanup of a login (logout): `SmokeTestFailure('cleanup.login failed')` -> `error fetching csrf token on /logout` -> `unexpected response code` with `{ actual, expected, url }` context.

> If we nevertheless want to wrap errors, it's possible to expose the info from the cause one level deep in a simple way without adding another method to OError. Instead of `OError.getFullInfoIncludeCause(err)`, you could write `{ OError.getFullInfo(err.cause), ...info }` or even: `{ ...info, cause: OError.getFullInfo(err.cause) }`. Or, you could do something similar in the `SmokeTestFailure` constructor and manually initialize the `info` object with the info from the underlying error. I think it would make it clearer that we're doing something different from the usual behaviour.

I need deep traversal, see the example above.

**Comment by emcsween** [comment]
It seems to me that the problem with ioredis reusing error objects is not specific to smoke tests. If we're worried about tagging errors in smoke tests, we should be worried about doing it anywhere. I think this problem should be dealt with at the interface where ioredis is used. Stuff that makes calls to redis ([here](https://github.com/overleaf/web-internal/blob/e5df4cde30d8b9e65062e1484699326e96c4eb92/app/src/Features/Compile/CompileManager.js#L201), for example) should wrap the error in an OError, so that we don't need to worry about tagging it later.

But I don't want to push back too much against having info objects propagate through causes. It's clear that the current behaviour is confusing, so I think we should simply make `getFullInfo()` do what `getFullInfoIncludeCause()` does and rely on other mechanisms for data that is not intended for logging.

> WDYT about Johns proposal of a custom OError class that allows nested context to get returned? (from OError.getFullInfo I assume)

I think what John means is that if we make `getFullInfo()` collect info objects in causes, we can still create custom errors that carry other bits of data that do not propagate through causes. For example, if you needed to create a "colored" error and you wanted to make sure that causes don't accidentally set a color on the main error, you could create a subclass of OError like this:
```js
class ColoredError extends OError {
  constructor(color, msg, info) {
    super(msg, info)
    this.color = color
  }
}
```

So, to recap, I think we should just merge what you've written into `getFullInfo()` and remove the distinction between tags and causes when it comes to how info is propagated. We ended up not making use of this distinction anyway.

File: index.js [hunk 1]
```diff
@@ -123,6 +123,25 @@ class OError extends Error {
...
+   * @param {Error | null | undefined} error any error (may or may not be an `OError`)
+   * @return {Object}
+   */
+  static getFullInfoIncludeCause(error) {
+    const info = OError.getFullInfo(error)
+    if (!error) return info
+
+    const oError = /** @type{OError} */ (error)
+    if (oError.cause) {
+      Object.assign(info, OError.getFullInfoIncludeCause(oError.cause))
```
Comment (emcsween) [inline index.js:140]: A little detail: written this way, if a property is present both in the main error and in a cause, the property value on the cause wins. We probably want to have it the other way around.

**Comment by das7pad** [comment]
> It seems to me that the problem with ioredis reusing error objects is not specific to smoke tests. If we're worried about tagging errors in smoke tests, we should be worried about doing it anywhere. I think this problem should be dealt with at the interface where ioredis is used. Stuff that makes calls to redis ([here](https://github.com/overleaf/web-internal/blob/e5df4cde30d8b9e65062e1484699326e96c4eb92/app/src/Features/Compile/CompileManager.js#L201), for example) should wrap the error in an OError, so that we don't need to worry about tagging it later.

Yes, we found out about it and dealt with it in at least one other service already: real-time. Maybe we should open a ticket for wrapping all the redis errors, in all services? That would be a lot of work tho.

> But I don't want to push back too much against having info objects propagate through causes. It's clear that the current behaviour is confusing, so I think we should simply make `getFullInfo()` do what `getFullInfoIncludeCause()` does and rely on other mechanisms for data that is not intended for logging.

I'll open a ticket for looking into the call-sites of `getFullInfo` and their consumed errors/causes, as noted at the last paragraph of https://github.com/overleaf/o-error/pull/14#issuecomment-734943950.

> > WDYT about Johns proposal of a custom OError class that allows nested context to get returned? (from OError.getFullInfo I assume)
> 
> I think what John means is that if we make `getFullInfo()` collect info objects in causes, we can still create custom errors that carry other bits of data that do not propagate through causes. For example, if you needed to create a "colored" error and you wanted to make sure that causes don't accidentally set a color on the main error, you could create a subclass of OError like this:

Oh, rereading his comment, he may have meant having a `err.publicInfo` separated from `err.info`. @jdleesmiller to elaborate on that.

File: index.js (hunk 1 above)
Comment (das7pad) [reply index.js:140]: I copied the behavior from `getFullInfo()`, which may get context details overwritten from tags.

**Comment by das7pad** [comment]
Investigation into possible exposure of OError context via causes completed in https://github.com/overleaf/issues/issues/3792 -- TLDR; no context will get exposed in merging the context of causes.

I will merge the new logic into `OError.getFullInfo` now.

**Comment by das7pad** [comment]
> I will merge the new logic into `OError.getFullInfo` now.

Done in https://github.com/overleaf/o-error/pull/14/commits/9f75060b21ae9e8176e0e4a8bacb7b69f9d1d281 and I also updated the PR description. We should squash merge this PR.

File: index.js [hunk 2]
```diff
@@ -114,6 +114,10 @@ class OError extends Error {
 
     if (typeof oError.info === 'object') Object.assign(info, oError.info)
 
+    if (oError.cause) {
+      Object.assign(info, OError.getFullInfo(oError.cause))
```
Comment (emcsween) [inline index.js:118]: To follow up on the previous discussion about the order of the merge, causes are slightly different from tags because they are lower in the stack. When you tag, you add context higher up the stack, so it makes sense that info from tags supersedes info from the original error. With causes, it's the opposite.

**Review by emcsween — COMMENTED** [review]
Great! Just a small detail on the order of the merge.

**Review by emcsween — APPROVED** [review]
Looks good to me! Thanks for indulging in the back and forth.

**Review by jdleesmiller — APPROVED** [review]
Thanks, both. Interesting discussion and investigation, and I like the result :)

On the ioredis problem: I think using `tag` and living with the possibility of confusing traces in some cases is not too bad, as an alternative to trying to wrap every redis callsite. I don't think we've seen what it looks like in production yet, so I'd be OK with trying it out at least (how often is it really a problem?).

On the `ColoredError` problem: this still feels to me like a bit of a rough edge in this library. On the one hand, putting a custom property into `info` is quick and easy, but it is not very type-friendly and blurs the line between debug info and info that the program can use and / or return to the user. On the other hand, a custom property is cleaner and avoids all those issues but won't get logged as `info`. I think there are some ways around this:
```js
// original
class ColoredError extends OError {
  constructor(color, msg, info) {
    super(msg, info)
    this.color = color
  }
}
let e1 = new ColoredError('blue', 'foo')
e1.color // blue
OError.getFullInfo(e1) // {}

// preserve the property but not the constructor
class ColoredError2 extends OError {
  get color() { return this.info.color }
}
let e2 = new ColoredError2('foo', { color: 'blue' })
e2.color // 'blue'
OError.getFullInfo(e2) // { color: 'blue' }

// with custom constructor and property
class ColoredError3 extends OError {
  constructor(color, msg, info) {
    super(msg, { ...info, color })
  }
  get color() { return this.info.color }
}
let e3 = new ColoredError3('blue', 'foo')
e3.color // 'blue'
OError.getFullInfo(e3) // { color: 'blue' }
```
With the (2) or (3) variants you get a clearer public interface for the error and the info still has everything for debug logging. WDYT? If you like it, we could document it as a suggested / recommended approach.

But, none of that is blocking here.

One other question: should we release this as 3.2 or 4.0 (coming from 3.1)? I guess 4.0 is probably the safer option. (Also: I realized I had not pushed the commit to bump the version to 3.1. I've now done that.)

**Comment by das7pad** [comment]
Thanks for elaborating on the custom error.

`ColoredError2.color` may throw in case no info object was passed. I would prefer `ColoredError3` as the color will end up in the log a well, for `ColoredError` it will only be visible in sentry.

On ioredis: I think we can leave the web usages as is then.

On the version: with a `3.x` release we can reinstall the logger module in web and it's a done deal. With a `4.x` release we need to bump o-error in the logger module and tag a release there as well, in order to get this patch into web. Hence I would favor a `3.x` release.
